### PR TITLE
Added support for python copy

### PR DESCRIPTION
**Pull request by ManishAradwad**
## Proposed changes

Please include a description of the problem or feature this PR is addressing. If there is a corresponding issue, include the issue #300 .

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by ManishAradwad** [comment]
@awni i have followed https://github.com/pytorch/pytorch/blob/035e55822ad123b02fbd9d91e1185f5c07af0ddd/torch/nn/modules/module.py#L1636-L1642 for the changes

Please let me know if I should change anything

File: python/mlx/nn/layers/base.py [hunk 1]
```diff
@@ -439,3 +439,11 @@ def _set_train(_, m):
 
     def eval(self):
         self.train(False)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop("_compiled_call_impl", None)
```
Comment (dc-dc-dc) [inline python/mlx/nn/layers/base.py:445]: What is `_compiled_call_impl` for ?

Comment (ManishAradwad) [reply python/mlx/nn/layers/base.py:445]: I was confused whether to add this line or not. I think torch doesn't need this particular value from state dict while copying. I'll remove this line.

Comment (dc-dc-dc) [reply python/mlx/nn/layers/base.py]: nit, can just return the copy instead of creating the variable first

**Comment by dc-dc-dc** [comment]
Can you add a basic test inside `python/tests/test_nn.py`

File: python/mlx/nn/layers/base.py (hunk 1 above)
Comment (ManishAradwad) [reply python/mlx/nn/layers/base.py]: done

Comment (ManishAradwad) [reply python/mlx/nn/layers/base.py]: @dc-dc-dc pls review

Comment (dc-dc-dc) [reply python/mlx/nn/layers/base.py]: LGTM, nice addition 🚀

File: ACKNOWLEDGMENTS.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ with a short description of your contribution(s) below. For example:
 
 MLX was developed with contributions from the following individuals:
 
+- Manish Aradwad: Added support for deepcopy.
```
Comment (awni) [inline ACKNOWLEDGMENTS.md:10]: Could you make this slightly more descriptive: Added support to pickle `nn.Module`?

**Review by awni — COMMENTED** [review]
Cool this is a nice simple change! Just a minor comment on the ack otherwise I think we can merge it.

File: ACKNOWLEDGMENTS.md (hunk 2 above)
Comment (ManishAradwad) [reply ACKNOWLEDGMENTS.md:10]: Sure!

**Review by awni — APPROVED** [review]
Thanks!

**Comment by awni** [comment]
PS thanks for the review @dc-dc-dc !

**Comment by awni** [comment]
This is breaking the CI 

> TypeError: cannot pickle 'mlx.core.array' object

Seems obvious in hindsight  but now I'm wondering how did that test pass for you? Or did you not run it?

**Comment by ManishAradwad** [comment]
My apologies, I assumed the changes were trivial so thought I can skip running the tests 🥲

I'll start working on the approach suggested by @dc-dc-dc

**Pull request merged**